alf/algorithms/off_policy_algorithm.py

```diff
@@ -147,139 +168,85 @@ def _train(self, experience, num_updates, mini_batch_size,
             policy_state = common.reset_state_if_necessary(
                 policy_state, initial_train_state,
-                tf.equal(exp.step_type, StepType.FIRST))
-
-            with tf.name_scope(scope):
-                policy_step = self.train_step(exp, policy_state)
-
-            info_ta = tf.nest.map_structure(
-                lambda ta, x: ta.write(counter, x), info_ta,
-                nest_utils.distributions_to_params(policy_step.info))
-
-            counter += 1
-
-            return [counter, policy_step.state, info_ta]
-
-        with tf.GradientTape(
-                persistent=True, watch_accessed_variables=False) as tape:
-            tape.watch(self.trainable_variables)
-            [_, _, info_ta] = tf.while_loop(
-                cond=lambda counter, *_: tf.less(counter, num_steps),
-                body=_train_loop_body,
-                loop_vars=[counter, first_train_state, info_ta],
-                back_prop=True,
-                name="train_loop")
-            info = tf.nest.map_structure(lambda ta: ta.stack(), info_ta)
-            info = nest_utils.params_to_distributions(
-                info, self.train_step_info_spec)
-            experience = nest_utils.params_to_distributions(
-                experience, self.processed_experience_spec)
-            training_info = TrainingInfo(
-                action=experience.action,
-                reward=experience.reward,
-                discount=experience.discount,
-                step_type=experience.step_type,
-                rollout_info=experience.rollout_info,
-                info=info,
-                env_id=experience.env_id)
-
-        loss_info, grads_and_vars = self.train_complete(
-            tape=tape, training_info=training_info, weight=weight)
-
-        del tape
-
-        return training_info, loss_info, grads_and_vars
+                exp.step_type == StepType.FIRST)
+            policy_step = self.train_step(exp, policy_state)
+            if self._train_info_spec is None:
+                self._train_info_spec = dist_utils.extract_spec(
+                    policy_step.info)
+            info_list.append(
+                dist_utils.distributions_to_params(policy_step.info))
+            policy_state = policy_step.state
+
+        info = alf.nest.utils.stack_nests(info_list)
+        info = dist_utils.params_to_distributions(info, self.train_info_spec)
+        experience = dist_utils.params_to_distributions(
+            experience, self.processed_experience_spec)
+        training_info = TrainingInfo(
+            action=experience.action,
+            reward=experience.reward,
+            discount=experience.discount,
+            step_type=experience.step_type,
+            rollout_info=experience.rollout_info,
+            info=info,
+            env_id=experience.env_id)
+
+        loss_info = self.calc_loss(training_info)
+        valid_masks = (training_info.step_type != StepType.LAST).to(
```

Comment:
FIRST step should still be used for loss, even though `prev_action` is not valid, the action is still used. And for most algorithm, `prev_action` is not used.

Reply:
The experience contains reward for prev_action, not for action, because experience is constructed using reward and prev_action from time_step and only action from alg_step.  https://github.com/HorizonRobotics/alf/blob/pytorch/alf/data_structures.py#L195

https://github.com/HorizonRobotics/alf/blob/pytorch/alf/data_structures.py#L91:
reward: a `Tensor` of reward values from executing 'prev_action'.

Is this still correct?

Reply:
> The experience contains reward for prev_action, not for action, because experience is constructed using reward and prev_action from time_step and only action from alg_step. https://github.com/HorizonRobotics/alf/blob/pytorch/alf/data_structures.py#L195
> 
> https://github.com/HorizonRobotics/alf/blob/pytorch/alf/data_structures.py#L91:
> reward: a `Tensor` of reward values from executing 'prev_action'.
> 
> Is this still correct?

The reward of FIRST won't be used for training, but you need observation of FIRST and reward of SECOND to update the value and actor networks given FIRST as the input. So we cannot mask out FIRST for training.

Reply:
I see, FIRST step rewards and discounted return are discarded here: https://github.com/HorizonRobotics/alf/blob/pytorch/alf/utils/value_ops.py#L159
`            acc_value = rets[t + 1] * discounts[t + 1] + rewards[t + 1]
            rets[t] = is_lasts[t] * values[t] + (1 - is_lasts[t]) * acc_value`
as long as the FIRST step is always following a LAST step.